src/libraries/System.Private.CoreLib/src/System/Threading/Wasi/WasiEventLoop.cs

```diff
@@ -29,18 +32,71 @@ internal static Task RegisterWasiPollable(Pollable pollable, CancellationToken c
             // this will register the pollable holder into s_pollables
             var holder = new PollableHolder(pollable, cancellationToken);
             s_pollables.Add(holder);
+
+            ScheduleCheck();
+
             return holder.taskCompletionSource.Task;
         }
 
-        // this is not thread safe
-        internal static void DispatchWasiEventLoop()
+
+        internal static T PollWasiEventLoopUntilResolved<T>(Task<T> mainTask)
+        {
+            s_mainTask = mainTask;
+            while (!mainTask.IsCompleted)
+            {
+                ThreadPoolWorkQueue.Dispatch();
+            }
+            s_mainTask = null;
+
+            var exception = mainTask.Exception;
+            if (exception is not null)
+            {
+                throw exception;
+            }
+
+            return mainTask.Result;
+        }
+
+        internal static void PollWasiEventLoopUntilResolvedVoid(Task mainTask)
         {
-            ThreadPoolWorkQueue.Dispatch();
-            if (s_tasksCanceled)
+            s_mainTask = mainTask;
+            while (!mainTask.IsCompleted)
+            {
+                ThreadPoolWorkQueue.Dispatch();
+            }
+            s_mainTask = null;
+
+            var exception = mainTask.Exception;
+            if (exception is not null)
             {
-                s_tasksCanceled = false;
-                return;
+                throw exception;
             }
+        }
+
+        internal static void CancelAllPollables()
```

Comment:
Is this function called anywhere?

Reply:
That was speculative cleanup code, to consider calling from `Main` but probably not from wasi exports. We can drop it.